File: questions/416.PartitionEqualSubsetSum.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
const canPartition = function (nums) {
  // Calculate sum of the array. If sum is odd, there can not be two subsets with equal sum, so return false.
  const sum = nums.reduce((acc, n) => acc + n, 0)
  if (sum % 2) return false

  let dp = new Set()
  dp.add(0)

  const target = sum / 2

  // If sum of array elements is even, store the sum of all elements using Dynamic Programming.
  for (let i = nums.length - 1; i >= 0; i--) {
    const tmp = new Set()
    dp.forEach((t) => {
      tmp.add(t + nums[i])
      tmp.add(t)
    })
    dp = tmp
  }

  return dp.has(target)
}

console.log(canPartition([1, 5, 11, 5]))
// console.log(canPartition([1,2,3,5]))

File: questions/416.PartitionEqualSubsetSum.go
package questions

func canPartition(nums []int) bool {
	sum := 0
	for _, num := range nums {
		sum += num
	}

	if sum%2 != 0 {
		return false
	}

	return canPartitionRecursive(map[int]map[int]bool{}, nums, sum/2, 0)
}

func canPartitionRecursive(dp map[int]map[int]bool, nums []int, sum, index int) bool {
	if sum == 0 {
		return true
	}

	if len(nums) == 0 {
		return false
	}
	if index >= len(nums) {
		return false
	}

	_, ok := dp[index]
	if !ok {
		dp[index] = map[int]bool{}
	}

	if _, ok := dp[index][sum]; !ok {
		// e.g. nums = [1,5,11,5], sum = 11
		// 1, 5, 5 is coming.
		if nums[index] <= sum {
			if canPartitionRecursive(dp, nums, sum-nums[index], index+1) {
				dp[index][sum] = true
				return true
			}
		}
	}

	// e.g. nums = [1,5,11,5], sum = 11
	// 11 is coming.
	dp[index][sum] = canPartitionRecursive(dp, nums, sum, index+1)

	return dp[index][sum]
}
